taichi/backends/cuda/cuda_profiler.cpp

```diff
@@ -60,44 +64,61 @@ void KernelProfilerCUDA::record(KernelProfilerBase::TaskHandle &task_handle,
-
-      if (Timelines::get_instance().get_enabled()) {
-        float time_since_base;
-        CUDADriver::get_instance().event_elapsed_time(&time_since_base,
-                                                      base_event_, start);
-        timeline.insert_event({map_elem.first, true,
-                               base_time_ + time_since_base * 1e-3, "cuda"});
-        timeline.insert_event(
-            {map_elem.first, false,
-             base_time_ + (time_since_base + kernel_time) * 1e-3, "cuda"});
-      }
 
-      auto it = std::find_if(
-          records.begin(), records.end(),
-          [&](KernelProfileRecord &r) { return r.name == map_elem.first; });
-      if (it == records.end()) {
-        records.emplace_back(map_elem.first);
-        it = std::prev(records.end());
-      }
-      it->insert_sample(kernel_time);
-      total_time_ms += kernel_time;
-
-      // TODO: the following two lines seem to increases profiler overhead a
-      // little bit. Is there a way to avoid the overhead while not creating
-      // too many events?
-      CUDADriver::get_instance().event_destroy(start);
-      CUDADriver::get_instance().event_destroy(stop);
+  // cuEvent : get traced kernel_elapsed_time
+  for (auto &record : event_records_) {
+    CUDADriver::get_instance().event_elapsed_time(
+        &record.kernel_elapsed_time_in_ms, record.start_event,
+        record.stop_event);
+    CUDADriver::get_instance().event_elapsed_time(
+        &record.time_since_base, base_event_, record.start_event);
+
+    // TODO: the following two lines seem to increases profiler overhead a
+    // little bit. Is there a way to avoid the overhead while not creating
+    // too many events?
+    CUDADriver::get_instance().event_destroy(record.start_event);
+    CUDADriver::get_instance().event_destroy(record.stop_event);
+
+    KernelProfileTracedRecord traced_record;
+    traced_record.name = record.name;
+    traced_record.kernel_elapsed_time_in_ms = record.kernel_elapsed_time_in_ms;
+    traced_record.time_since_base = record.time_since_base;
+    traced_records_.push_back(traced_record);
+  }
+
+  // statistics on traced_records_
+  for (auto &record : traced_records_) {
+    auto it =
+        std::find_if(statistical_results_.begin(), statistical_results_.end(),
+                     [&](KernelProfileStatisticalResult &result) {
+                       return result.name == record.name;
+                     });
+    if (it == statistical_results_.end()) {
+      statistical_results_.emplace_back(record.name);
+      it = std::prev(statistical_results_.end());
     }
+    it->insert_record(record.kernel_elapsed_time_in_ms);
+    total_time_ms_ += record.kernel_elapsed_time_in_ms;
   }
-  outstanding_events_.clear();
+
+  // timeline
+  if (Timelines::get_instance().get_enabled()) {
+    auto &timeline = Timeline::get_this_thread_instance();
+    for (auto &record : traced_records_) {
+      timeline.insert_event({record.name, true,
```

Comment:
nit: Put a `/*param_name=*/` before these literal values so that ppl know what they are for :-)